File: src/main/java/net/dumbdogdiner/dogcore/commands/TellCommand.java
package net.dumbdogdiner.dogcore.commands;

import dev.jorel.commandapi.annotations.Alias;
import dev.jorel.commandapi.annotations.Command;
import dev.jorel.commandapi.annotations.Default;
import dev.jorel.commandapi.annotations.arguments.AEntitySelectorArgument;
import dev.jorel.commandapi.annotations.arguments.AGreedyStringArgument;
import java.util.Collection;
import net.dumbdogdiner.dogcore.chat.MiscFormatter;
import net.dumbdogdiner.dogcore.database.User;
import net.dumbdogdiner.dogcore.messages.Messages;
import net.kyori.adventure.text.Component;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

@Command("tell")
@Alias({"msg", "w", "whisper", "pm", "t"})
public final class TellCommand {
    private TellCommand() { }

    @Default
    public static void tell(
        final @NotNull CommandSender sender,
        final @NotNull @AEntitySelectorArgument.ManyPlayers Collection<Player> players,
        final @NotNull @AGreedyStringArgument String message
    ) {
        User.nameIfNotMuted(sender).thenAccept(name -> {
            if (name != null) {
                var receiverName = MiscFormatter.playerCollection(players);

                var messageComponent = Component.text(message);
                sender.sendMessage(Messages.get("chat.tell.outgoing", receiverName, messageComponent));
                for (var player : players) {
                    player.sendMessage(Messages.get("chat.tell.incoming", name, messageComponent));
                }

                User.spies().thenAccept(spies -> {
                    if (!spies.isEmpty()) {
                        var spyMessage = Messages.get("chat.tell.spy", name, receiverName, messageComponent);

                        for (var spy : spies) {
                            spy.sendMessage(spyMessage);
                        }
                    }
                });
            }
        });
    }
}

File: src/main/kotlin/net/dumbdogdiner/dogcore/commands/TellCommand.kt
package net.dumbdogdiner.dogcore.commands

import net.dumbdogdiner.dogcore.chat.NameFormatter
import net.dumbdogdiner.dogcore.database.User
import net.dumbdogdiner.dogcore.messages.Messages
import net.kyori.adventure.text.Component
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import revxrsal.commands.annotation.Command

object TellCommand {
    @Command("tell", "msg", "w", "whisper", "pm", "t")
    fun tell(sender: CommandSender, receiver: Player, message: String) {
        // check if muted
        val senderName = if (sender is Player) {
            if (User.lookup(sender)?.isMuted == true) {
                commandError(Messages["error.muted"])
            }
            NameFormatter.formatUsername(sender).get()
        } else {
            sender.name()
        }
        val receiverName = NameFormatter.formatUsername(receiver).get()

        val messageComponent = Component.text(message)
        sender.sendMessage(Messages["chat.tell.outgoing", receiverName, messageComponent])
        receiver.sendMessage(Messages["chat.tell.incoming", senderName, messageComponent])

        val spies = User.spies()
        if (spies.isNotEmpty()) {
            val spyMessage = Messages["chat.tell.spy", senderName, receiverName, messageComponent]

            for (spy in spies) {
                spy.sendMessage(spyMessage)
            }
        }
    }
}
